src/strings/gb18030.c

```diff
@@ -0,0 +1,346 @@
+#include "moar.h"
+#include "gb18030_codeindex.h"
+
+/*  Information about GB18030: http://www.gb18030.com/
+    GB18030 to Unicode Mapping used (included in GNU LIBICONV Package):
+	ftp://ftp.gnu.org/pub/gnu/libiconv/libiconv-1.7.tar.gz
+*/
+
+const MVMint32 gb18030_two_byte_lower_bound[126] = {
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
+64,64,64,64,64,64,161,161,161,161,161,161,161,64,64,64,64,64,64,64,64,64,64,64,
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,
+64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64,64};
+
+const MVMint32 gb18030_two_byte_upper_bound[126] = {
+254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,254,254,254,254,254,254,254,252,254,243,246,245,241,233,
+239,160,160,160,160,160,160,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,249,254,254,254,254,254,254,254,254,254,254,254,254,254,
+254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,254,160,
+160,160,160,160,160,159};
+
+MVMint32 gb18030_valid_check_len2(MVMint32 c_1, MVMint32 c_2) {
+    /* This function serves like a 'first stage check' of c_1 and c_2.
+       It eliminates most of the invalid combinations of c_1 and c_2, 
+       but for code simplicity and to avoid lots of if-else here,
+       ther remaining invalid combinations will be processed 
+       in the conversion table. */
+    if (c_1 < 0x81 || c_1 > 0xfe) return 0;
+    c_1 -= 0x81;
+    return gb18030_two_byte_lower_bound[c_1] <= c_2 && c_2 <= gb18030_two_byte_upper_bound[c_1] ? 1 : 0;
+}
+
+MVMint32 gb18030_valid_check_len4(MVMint32 c_1, MVMint32 c_2, MVMint32 c_3, MVMint32 c_4) {
+    if ((0x81 <= c_1 && c_1 <= 0x83) || (c_1 == 0x84 && c_2 == 0x30)) {
+        return (0x30 <= c_2 && c_2 <= 0x39) && (0x81 <= c_3 && c_3 <= 0xfe) && (0x30 <= c_4 && c_4 <= 0x39) ? 1 : 0;
```

Comment:
The `? :` ternary operator is not needed as `(0x30 <= c_2 && c_2 <= 0x39) && (0x81 <= c_3 && c_3 <= 0xfe) && (0x30 <= c_4 && c_4 <= 0x39)` will evaluate to if true to 1 or false as 0 without it.